<html>

<head>
<title>Crawl2 - Welcome</title>
<link rel="stylesheet" href="http://www.genedb.org/includes/style/genedb/main.css" type="text/css" />

<script type="text/javascript" src="http://www.genedb.org/includes/scripts/jquery/jquery-genePage-combined-new.js"></script>
<script src="http://samaxesjs.googlecode.com/files/jquery.toc-1.1.1.js"></script>

<style>

body {
    font-size: 1em;
    width:900px;
    margin:auto;
}

h1 { font-size:2em; }
h2 { font-size:1.8em; }
h3 { font-size:1.6em; }
h4 { font-size:1.4em; }
h5 { font-size:1.2em; }

h1,h2,h3,h4,h5 {
    font-weight:bold;
    text-transform: uppercase;
}

p {
    padding:15px;
    text-align: justify;
}

pre {
    margin:10px;
    padding:10px;
    border: 1px dashed grey;
}

table {
    margin:20px;
    
}

table, td, th {
    border: 1px dashed grey;
}

td, th {
    padding:10px;
}

div#toc {
    border: 1px dashed grey;
    padding:10px;
    width:400px;
    margin:auto;
}

div#toc ul {
    list-style: none;
    text-transform: uppercase;
    font-size:small;
}

div#toc ul li ul {
    margin-bottom: 0.75em;
}

div#toc ul li ul li ul {
    margin-bottom: 0.25em;
}

a, a:visited {
    text-decoration:none;
    color:grey;
}

a:hover {
    color:black;
}

li {
    margin-left:4em;
}

</style>

<link rel="icon" href="crawl-discpng.png" type="image/x-icon" />  


</head>
<body>



<div style="width:800px;margin:auto;" >
<img src="calligraphic.png">
</div>


<div style="float:right;margin:20px;">
<div id="toc"></div>
<P>
    <em>
        Note: Generated using <a href="http://jquery.org/">Jquery</a> and a <a href="http://code.google.com/p/samaxesjs/wiki/TableOfContentsPlugin">TOC plugin</a>.
    </em>
</P>
</div>

<h1 id="welcome">Welcome</h1>

<H2 id="setup">Setup guide</H2>

<ul><li><a href="setup.html">Setup Guide</a></li></ul>

<h2 id="web-artemis">Visually explore the data</h2>

<P><a href="https://github.com/sanger-pathogens/Web-Artemis">Web-Artemis</a> is a <em>very cool</em> genome browser. It can talk
to Crawl services, so can be used to explore the data they hold. 

<P>If it has been <a href="setup.html#webartemis">pulled</a> into this web app, you can browse the data with the Very Nifty :
<ul><li><a href="./Web-Artemis/index.html">Web Artemis</a></li></ul>

<P>(Or you can install it yourself by grabbing it : </P>

<code><pre>git clone https://github.com/sanger-pathogens/Web-Artemis.git</pre></code>

<P>and setting it up on this or another web-server.</P>



<h2 id="services">REST Services</h2>

<P>Metadata about each service can be accessed at : </P>

<ul><li><a href="./organisms/">Organisms</a></li>
<li><a href="./regions/">Regions</a></li>
<li><a href="./features/">Features</a></li>
<li><a href="./organisms/">SAM/BAM files</a></li>
<li><a href="./variants/">VCF/BCF files</a></li>
</ul>

<h3 id="organisms">Organisms</h3>

<ul>
    <li><a href="./organisms/list">Listing all organisms</a></li>
    <li><a href="./organisms/get?organism=com:Pfalciparum">Getting Pfalciparum</a></li>
</ul>

<h3 id="regions">Regions</h3>

<ul>
    
    <li><a href="./regions/inorganism.json?organism=org%3A27">Getting regions in an oranism.</a></li>    
    <li><a href="./regions/sequence.json?region=Pf3D7_01&start=1&end=4001">Getting the sequence of a region</a></li>
    <li><a href="./regions/locations.json?&region=Pf3D7_01&start=1&end=4001&types=gene&types=pseudogene&exclude=false">Getting features on a sequence.</a></li>
    
</ul>

<h3 id="features">Features</h3>

<ul>
    <li><a href="./features/hierarchy.json?features=PFA0005w%3Aexon%3A1&root_on_genes=true">Getting the gene hiearchy of a feature</a></li>
    <li><a href="./features/synonyms.json?features=PFA0005w:exon:1&features=PFA0005w&features=PFA0005w:mRNA&features=PFA0005w:pep">Getting synonyms</a></li>
    <li><a href="./features/properties.json?features=PFA0005w:exon:1,PFA0005w,PFA0005w:pep">Getting properties</a></li>
    <li><a href="./features/pubs.json?features=PFA0005w:exon:1&features=PFA0005w&features=PFA0005w:mRNA&features=PFA0005w:pep">Getting pubs</a></li>
    <li><a href="./features/dbxrefs.json?features=PFA0005w:exon:1&features=PFA0005w&features=PFA0005w:mRNA&features=PFA0005w:pep">Getting dbxrefs</a></li>
    <li><a href="./features/terms.json?features=PFA0005w:exon:1&features=PFA0005w&features=PFA0005w:mRNA&features=PFA0005w:pep">Getting cv terms</a></li>
    <li><a href="./features/orthologues.json?features=PFA0005w:exon:1&features=PFA0005w&features=PFA0005w:mRNA&features=PFA0005w:pep">Getting orthologues</a></li>
    
    
</ul>

<h3 id="bam">Sequence Alignment/Map files (SAM/BAM)</h3>

<ul>
    
    <li><a href="./sams/listforsequence.json?&callback=jsonp1304795281789&sequence=Pf3D7_01">Listing alignment files that match a sequence</a></li>
    <li><a href="./sams/query.json?&fileID=1&sequence=Pf3D7_01&start=1&end=4001&filter=4&properties=alignmentBlocks&properties=readName&properties=flags">Querying an alignment</a></li>
    <li><a href="./sams/coverage.json?fileID=4&sequence=Pf3D7_01&start=1&end=5001&window=13">Getting the coverage for an alignment</a></li>
</ul>

<h3 id="vcf">Variation files (VCF/BCF)</h3>

<ul>
    <li><a href="./variants/list">Listing all variant files</a></li>
    <li><a href="./variants/listfororganism?organism=com:Spneumoniae">Listing all variant files that match an organism</a></li>
    <li><a href="./variants/listforsequence?sequence=S_pneumoniae_Spanish_23F.dna">Listing all variant files that match a region in an organism</a></li>
    <li><a href="./variants/sequences?fileID=0">Listing all regions in a variant file</a></li>
    <li><a href="./variants/query.json?fileID=0&sequence=S_pneumoniae_Spanish_23F.dna&start=1&end=1000000">Querying a variant file</a></li>
    <li><a href="./variants/query.json?fileID=0&sequence=S_pneumoniae_Spanish_23F.dna&start=1&end=1000000&filter=3">Querying a variant file, using an integer filter.</a></li>
    <li><a href="./variants/queryWithFilters?fileID=0&sequence=S_pneumoniae_Spanish_23F.dna&start=1&end=10000000&filters=SHOW_SYNONYMOUS&filters=SHOW_NON_SYNONYMOUS&filters=SHOW_DELETIONS&filters=SHOW_INSERTIONS&filters=SHOW_MULTI_ALLELES&filters=SHOW_NON_OVERLAPPINGS&filters=SHOW_NON_VARIANTS&filters=MARK_NEW_STOPS">Querying a variant file, using string filters.</a></li>
</ul>

<h2 id="soap services">SOAP Services</h2>

<P>As an alternative to the RESTful approach, Crawl also publishes SOAP services too.</P>

<P>These <a title="SOAP services" href="soap/"> services, and their WSDLS, are described here</a> in more detail. </P>

<P>All the same Crawl RESTful resources are available as SOAP endpoints too. What follows is a simple python program illustrating how to use them.</P> 

<pre><code>
# first you must "easy_install suds"

import suds
from suds.client import Client

wsdl = "http://localhost:8080/services/soap/%s?wsdl"

organisms = Client(wsdl % "organisms")
regions = Client(wsdl % "regions")
features = Client(wsdl % "features")

print organisms
print regions
print features


first_organism = organisms.service.listOrganisms()[0]
print first_organism

# get the Pfalciparum organism
pfalciparum = organisms.service.getByString('Pfalciparum')[0]
print pfalciparum

# get the first region in Pfalciparum
first_pfalciparum_region = regions.service.inorganism(pfalciparum._commonName)[0]
print first_pfalciparum_region

# extract the first gene from all the regions in Pfalciparum
for region in regions.service.inorganism(pfalciparum._commonName):
    genes = regions.service.locations(region._uniqueName, 1, 1000)
    if len(genes) > 0 :
        gene = genes[0]
        print gene
        print features.service.hierarchy(genes[0]._uniqueName )



</code></pre>


<script>

$(document).ready(function() {
    $('#toc').toc();
});

</script>



</body>

</html>